src/machine/machine_stm32_spi.go

```diff
@@ -17,83 +39,178 @@ type SPIConfig struct {
 	SDI       Pin
 	LSBFirst  bool
 	Mode      uint8
+	Direction uint8
+	DataSize  uint8
+	CSMode    uint8
```

Comment:
Can you please explain the semantics of all these?

* What is the impact of the `Direction` flag on the Tx and Transfer functions?
* What impact has `DataSize`, considering that `Transfer` and `Tx` only support byte slices?
* How is the CS pin configured?

While there are other chips with such features, they aren't implemented for TinyGo. And I definitely think such things could be useful, but they must make sense in the provided API. For example, instead of setting a "soft" or "hard" CS pin, you could pass in the pin itself (or `NoPin`).

Right now I think all these features are already more or less supported in TinyGo with software. Half duplex communication is implemented by leaving either the `w` or the `r` parameter of `Tx` empty, the data size is not relevant on the wire (it's all just bits), and CS is easily emulated in software. In other words, I suspect these are performance optimizations instead of supporting something you could not do otherwise.

All in all I think it would be better to leave them unimplemented for now until one (or multiple!) use cases appear and to model an API based on that. The reason is that it is very easy to add an API but very hard to remove or change it later.

Reply:
Well if you'd rather I just remove them for now until other machines come up with a common API, I can do that -without scouring back through the reference manual to try (and probably fail) to define their behaviors.

The one difference that might affect bus stability is clearing the overrun flag for full-duplex mode (lines 206-209). If we default to this mode, and the user implements a half or simplex mode in software, then we could be inadvertently clearing important error flags.

The motivation for clearing this flag in this mode comes from ST's own SPI driver source code.

Reply:
Hi @ardnew I think removing is the correct answer, at lest for now. Thanks!

Reply:
Yes, indeed. The reason is that APIs (especially one as fundamental as the machine package) should have small, well  defined interfaces.